BUG: Fixed file handle leak in array_tofile.

simon-graham opened the pull request:
Closes #17589

If the ```PyArray_ToFile``` call inside ```array_tofile``` fails for any reason, we leak a dup-ed file handle. This causes problems especially on Windows OS, where any subsequent attempt to delete the file will fail because NumPy is still holding an open handle.

This update ensures that we always close the handle.

The added test case forces ```PyArray_ToFile``` to fail, which triggers the original problem.

simon-graham committed:
BUG: Fixed file handle leak in array_tofile.

The dup-ed file handle created in array_tofile is now closed when PyArray_ToFile fails.

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -619,10 +619,9 @@ array_tofile(PyArrayObject *self, PyObject *args, PyObject *kwds)
+    if (res1 < 0 || res2 < 0) {
         goto fail;
     }
     if (own && npy_PyFile_CloseFile(file) < 0) {
```
if `res2 < 0`, you still should close this file

simon-graham replied:
Thanks @eric-wieser. I've updated it now so that it closes ```file``` even if ```npy_PyFile_DupClose2``` fails.

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -619,10 +619,9 @@ array_tofile(PyArrayObject *self, PyObject *args, PyObject *kwds)
-        goto fail;
-    }
-    if (npy_PyFile_DupClose2(file, fd, orig_pos) < 0) {
+    res1 = PyArray_ToFile(self, fd, sep, format);
+    res2 = npy_PyFile_DupClose2(file, fd, orig_pos);
+    if (res1 < 0 || res2 < 0) {
         goto fail;
     }
```
I suspect this might be unsafe - if the first throws an exception, it's not safe to call back into python code before catching it - and `npy_PyFile_DupClose2` looks like it can call back in. You might need to use
```suggestion
    int res_write = PyArray_ToFile(self, fd, sep, format);
    {
        PyObject *type, *value, *traceback;
        PyErr_Fetch(&type, &value, &traceback);
        int res_close = npy_PyFile_DupClose2(file, fd, orig_pos);
        npy_PyErr_ChainExceptions(type, value, traceback);
        if (res_write < 0 || res_close < 0) {
            return NULL;
        }
    }
```
which also means you'll get chained exceptions

simon-graham committed:
Update numpy/core/src/multiarray/methods.c

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

simon-graham committed:
BUG: array_tofile now always closes the original file handle.

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -585,8 +585,9 @@ array_tostring(PyArrayObject *self, PyObject *args, PyObject *kwds)
 array_tofile(PyArrayObject *self, PyObject *args, PyObject *kwds)
 {
-    int own;
+    int own, res_write, res_close;
     PyObject *file;
+    PyObject *err_type = NULL, *err_value = NULL, *err_traceback = NULL;
```
We use C99, so I'd prefer it if you put these declarations at their point of use:

```suggestion
    int own;
    PyObject *file;
```

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -617,23 +618,29 @@ array_tofile(PyArrayObject *self, PyObject *args, PyObject *kwds)
     }
-    if (npy_PyFile_DupClose2(file, fd, orig_pos) < 0) {
-        goto fail;
+    res_write = PyArray_ToFile(self, fd, sep, format);
+    PyErr_Fetch(&err_type, &err_value, &err_traceback);
+
+    res_close = npy_PyFile_DupClose2(file, fd, orig_pos);
+    if (res_close < 0) {
+        npy_PyErr_ChainExceptions(err_type, err_value, err_traceback);
```
```suggestion
    int res_write = PyArray_ToFile(self, fd, sep, format);
    PyObject *err_type, *err_value, *err_traceback;
    PyErr_Fetch(&err_type, &err_value, &err_traceback);

    int res_close = npy_PyFile_DupClose2(file, fd, orig_pos);
    if (res_close < 0) {
        npy_PyErr_ChainExceptions(err_type, err_value, err_traceback);
```
There's also no need to initialize the `err_*` variables, `PyErr_Fetch` is guaranteed to write to them.

eric-wieser commented:
Thanks for the patch. I think the error handling is still generally wrong in this function. I think we need to aim for the C equivalent of:

```python
file = ...
try:
    fd = npy_PyFile_Dup2(file)
    try:
        PyArray_ToFile(fd)
    finally:
        npy_PyFile_DupClose2(fd)
finally:
    if own:
        npy_PyFile_CloseFile(file)
```
This would almost certainly be much easier to get right if the contents of the outer `try` were extracted to a helper function. A `finally` clause is a `PyErr_Fetch` followed by an _unconditional_ call to `npy_PyErr_ChainExceptions`.

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -617,23 +617,31 @@ array_tofile(PyArrayObject *self, PyObject *args, PyObject *kwds)
+        npy_PyErr_ChainExceptions(err_type, err_value, err_traceback);
     }
+
     if (own && npy_PyFile_CloseFile(file) < 0) {
```
I think `npy_PyFile_CloseFile` is also unsafe to call here - you need to fetch the exception _again_, in case `npy_PyFile_DupClose2` created one. See my comment in https://github.com/numpy/numpy/pull/17598#issuecomment-713902244, this becomes much easier to follow if you extract a helper function.

simon-graham replied:
Thanks @eric-wieser, appreciate your patience. I didn't quite manage to refactor it into a helper function, but have updated it so that is now hopefully exception safe, and not too horrific to read.

simon-graham committed:
BUG: Made array_tofile exception safe.

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -616,24 +616,36 @@ array_tofile(PyArrayObject *self, PyObject *args, PyObject *kwds)
-        goto fail;
-    }
-    if (npy_PyFile_DupClose2(file, fd, orig_pos) < 0) {
-        goto fail;
+    PyObject *err_type, *err_value, *err_traceback;
+    int failed = 0;
+
+    if (fd != NULL) {
+        if (PyArray_ToFile(self, fd, sep, format) < 0) {
+            failed = 1;
+        }
+        PyErr_Fetch(&err_type, &err_value, &err_traceback);
+        if (npy_PyFile_DupClose2(file, fd, orig_pos) < 0) {
+            failed = 1;
+        }
+        npy_PyErr_ChainExceptions(err_type, err_value, err_traceback);
+    } else {
+        failed = 1;
     }
```
Would you mind extracting these to a helper function? Perhaps
```C
/* Like PyArray_ToFile but takes the file as a python object */
static int
PyArray_ToFileObject(PyArrayObject *self, PyObject *file, char *sep, char *sep)
{
    npy_off_t orig_pos = 0;
    FILE *fd = npy_PyFile_Dup2(file, "wb", &orig_pos);
    if (fd == NULL) {
        return -1;
    }
    int write_ret = PyArray_ToFile(self, fd, sep, format);
    PyObject *err_type, *err_value, *err_traceback;
    PyErr_Fetch(&err_type, &err_value, &err_traceback);
    int close_ret = npy_PyFile_DupClose2(file, fd, orig_pos);
    npy_PyErr_ChainExceptions(err_type, err_value, err_traceback);
    if (write_ret || close_ret) {
        return -1;
    }
    return 0;
}
```

simon-graham committed:
MAINT: Refactored array_tofile to use a helper function.

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -5048,6 +5048,14 @@ def test_tofile_format(self):
             s = f.read()
         assert_equal(s, '1.51,2.00,3.51,4.00')
 
+    def test_tofile_cleanup(self):
+        x = np.zeros((10), dtype=object)
+        with open(self.filename, 'wb') as f:
+            assert_raises(IOError, lambda: x.tofile(f, sep=''))
+
+        # Dup-ed file handle should be closed or remove will fail.
+        os.remove(self.filename)
+
```
Can you repeat this test with `x.tofile(self.filename)`, to check it closes both the python handle and the fd?

simon-graham replied:
This second check is now added.

simon-graham committed:
TST: Check that tofile closes its Python handle when we pass it a str.

If we pass a str to tofile() rather than an open file object, it will open a Python handle. Make sure that it gets closed correctly.

simon-graham commented:
Hi @eric-wieser, is there anything else to address for this PR? I believe I've implemented all of your review feedback. Thanks.

eric-wieser reviewed (approved):
Sorry for forgetting about this - looks good, thanks!

mattip commented:
Thanks @simon-graham
